chain/epoch_manager/src/tests/mod.rs

```diff
@@ -2442,3 +2451,208 @@ fn test_chunk_producers() {
+            ("test2".parse().unwrap(), NotEnoughChunks { produced: 70, expected: 100 }),
+            ("test3".parse().unwrap(), NotEnoughBlocks { produced: 89, expected: 100 }),
+        ])
+    );
+    assert_eq!(
+        validator_stats,
+        HashMap::from([
+            (
+                "test0".parse().unwrap(),
+                BlockChunkValidatorStats {
+                    block_stats: ValidatorStats { produced: 100, expected: 100 },
+                    chunk_stats: ValidatorStats { produced: 170, expected: 200 }
+                }
+            ),
+            (
+                "test4".parse().unwrap(),
+                BlockChunkValidatorStats {
+                    block_stats: ValidatorStats { produced: 0, expected: 0 },
+                    chunk_stats: ValidatorStats { produced: 100, expected: 100 }
+                }
+            ),
+        ])
+    );
+}
+
+#[test]
+/// Test that the stake of validators kicked out in an epoch doesn't exceed the max_kickout_stake_ratio
+fn test_max_kickout_stake_ratio() {
+    #[allow(unused_mut)]
+    let mut epoch_config =
+        epoch_config(5, 2, 4, 0, 90, 80, 0, None).for_protocol_version(PROTOCOL_VERSION).clone();
+    // At most 50% of total stake can be kicked out
+    #[cfg(feature = "protocol_feature_max_kickout_stake")]
+    {
+        epoch_config.validator_max_kickout_stake_perc = 50;
+    }
+    let accounts = vec![
+        ("test0".parse().unwrap(), 1000),
+        ("test1".parse().unwrap(), 1000),
+        ("test2".parse().unwrap(), 1000),
+        ("test3".parse().unwrap(), 1000),
+        ("test4".parse().unwrap(), 500),
+    ];
+    let epoch_info = epoch_info(
+        0,
+        accounts,
+        vec![0, 1, 2, 3],
+        vec![vec![0, 1], vec![2, 4]],
+        vec![],
+        vec![],
+        BTreeMap::new(),
+        vec![],
+        HashMap::new(),
+        0,
+    );
+    let (kickouts, validator_stats) = EpochManager::compute_kickout_info(
+        &epoch_config,
+        &epoch_info,
+        &HashMap::from([
+            (0, ValidatorStats { produced: 50, expected: 100 }),
+            (1, ValidatorStats { produced: 60, expected: 100 }),
+            (2, ValidatorStats { produced: 70, expected: 100 }),
+            // validator 3 doesn't need to produce any block or chunk
+            (3, ValidatorStats { produced: 0, expected: 0 }),
+        ]),
+        &HashMap::from([
+            (
+                0,
+                HashMap::from([
+                    (0, ValidatorStats { produced: 0, expected: 100 }),
+                    (1, ValidatorStats { produced: 0, expected: 100 }),
+                ]),
+            ),
+            (
+                1,
+                HashMap::from([
+                    (2, ValidatorStats { produced: 100, expected: 100 }),
+                    (4, ValidatorStats { produced: 50, expected: 100 }),
+                ]),
+            ),
+        ]),
+        &HashMap::new(),
+        &HashMap::new(),
```

Comment:
can you add one test with prev_validators kickout not-empty?

Reply:
There are some existing tests in the codebase that's for testing not all validators are kicked out in two epochs, for example, `test_validator_kickout`, so I didn't add a test specifically for this.